doc/dev/docstring_typehint.md

```diff
@@ -0,0 +1,185 @@
+```python
+:param <type> <param_name>: <Description of the parameter>
+```
+2. This option works best for parameters that use custom models or have many different options
+```python
+:param <param_name>: <Description of the parameter>
+:type <param_name>: <param_type>
+```
+
+Optional keyword arguments can in one-line or two-lines. Both options can be used in a single docstring for documenting different keywords without issue.
+1. This option works best for keyword args that are one of the basic types (`str`, `int`, `bool`, `bytes`, `float`, etc.)
+```python
+:keyword <type> <keyword_name>: <Description of the keyword>
+```
+2. This option works best for keyword that use custom models or have many different options
+```python
+:keyword <keyword_name>: <Description of the keyword>
+:paramtype <keyword_name>: <keyword_type>
+```
+
+The returned object is documented on two lines, the first describing what the returned object is and the second describing what the returned type is.
+```python
+"""
+:return: <Description of the returned object>
+:rtype: <Type of the returned object>
+"""
+```
+
+Finally, describe the possible errors raised by a method:
+```python
+:raises <error1>, <error2>, or <error3>: <description>
+```
+
+All of the above information needs to be added for each public method and there **needs to be a newline after the above docstrings**. There are additional options for including examples and version specific additions.
+
+For adding a version specific change use the `versionadded` docstring:
+```
+.. versionadded:: <version_number>
+
+```
+
+### Model and Client Docstrings
+
+For documenting the properties of a model or client, include a docstring right after the `__init__` method or immediately after the class declaration
+
+Here is an example in `azure-ai-textanalytics`:
+```python
+class DetectedLanguage(DictMixin):
+    """DetectedLanguage contains the predicted language found in text,
+    its confidence score, and its ISO 639-1 representation.
+
+    :ivar name: Long name of a detected language (e.g. English,
+        French).
+    :vartype name: str
+    :ivar iso6391_name: A two letter representation of the detected
+        language according to the ISO 639-1 standard (e.g. en, fr).
+    :vartype iso6391_name: str
+    :ivar confidence_score: A confidence score between 0 and 1. Scores close
+        to 1 indicate 100% certainty that the identified language is true.
+    :vartype confidence_score: float
+    """
+```
+
+The properties of a model should be documented with the `ivar` docstring:
+```
+:ivar <property_name>: <description>
+:vartype <property_name>: <type>
+```
+
+## Type Hints
+
+### Type Hints for Python 2.7 and 3.5+
+
+Python 2.7 does not support in-line type hints, the type hints for all sync code will have to adhere to type-hints in comments only. On all public methods include a type-hint by using the `# type: (...) -> (...)` format:
+```python
+def add(num1, num2):
+    # type: (int, int) -> int
+    ...
```

Comment:
maybe show in the snippet that `add()` is returning an `int` so the return type comment makes sense